### PR TITLE
Issue 19006 cleanupfieldreferencesjob is not stateful

### PR DESCRIPTION
**Pull request by nollymar**
With these changes we are forcing stateful jobs to create only one trigger instance when multiple jobs need to be created dynamically

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/quartz/DotJobStore.java [hunk 1]
```diff
@@ -34,7 +36,22 @@
 
 	@Override
 	public void storeJobAndTrigger(SchedulingContext ctxt, JobDetail newJob, Trigger newTrigger) throws ObjectAlreadyExistsException, JobPersistenceException {
-		super.storeJobAndTrigger(ctxt, newJob, newTrigger);
+
+        if (StatefulJob.class.isAssignableFrom(newJob.getJobClass())) {
+            this.executeInLock(this.isLockOnInsert() ? "TRIGGER_ACCESS" : null, conn -> {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/quartz/DotJobStore.java:41]: Codacy found an issue: [Assigning an Object to null is a code smell.  Consider refactoring.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+ * jobs
+ *
+ * @author nollymar
+ */
+public class DotStatefulJobTest {
+
+    private final static int THREAD_AMOUNT = 7;
+
+    @BeforeClass
+    public static void prepare() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java:34]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+    /**
+     * Inner class with the logic to be executed by each thread in {@link
+     * DotStatefulJobTest#testDotStatefulJobIsReallyStateful()} ()}
+     */
+    private class StatefulTestThread implements Runnable {
+
+        private CyclicBarrier cyclicBarrier;
+        private CountDownLatch countDownLatch;
+        private Scheduler scheduler;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java:106]: Codacy found an issue: [Private field 'scheduler' could be made final; it is only initialized in the declaration or constructor.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.TimeUnit;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.quartz.JobDataMap;
+import org.quartz.JobDetail;
+import org.quartz.Scheduler;
+import org.quartz.SchedulerException;
+import org.quartz.Trigger;
+import org.quartz.core.SchedulingContext;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java:21]: Codacy found an issue: [Avoid unused imports such as 'org.quartz.core.SchedulingContext'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                final JobDetail jobDetail = getJobDetail(thisThreadName);
+
+                final Trigger trigger = new TriggerBuilder().jobDetail(jobDetail)
+                        .triggerGroupName("StatefulTestJob_trigger_group").build();
+
+                //wait for all threads to schedule jobs concurrently
+                cyclicBarrier.await();
+                scheduler.scheduleJob(jobDetail, trigger);
+                countDownLatch.countDown();
+            } catch (Exception e) {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java:130]: Codacy found an issue: [Avoid catching generic exceptions such as NullPointerException, RuntimeException, Exception in try-catch block](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                throw new DotRuntimeException(e);
+            }
+        }
+    }
+
+    /**
+     * Method used to create a job detail instance
+     */
+    private JobDetail getJobDetail(String thisThreadName) {
+        System.out.println("Triggering job for thread " + thisThreadName);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java:177]: Codacy found an issue: [System.out.println is used](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/main/java/com/dotmarketing/quartz/DotStatefulJob.java [hunk 7]
```diff
@@ -1,7 +1,31 @@
...
+import java.util.Date;
+import org.quartz.JobDetail;
+import org.quartz.SimpleTrigger;
 import org.quartz.StatefulJob;
+import org.quartz.Trigger;
 
 public abstract class DotStatefulJob extends DotJob implements StatefulJob {
 
+    public static final class TriggerBuilder {
+        private JobDetail jobDetail;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/quartz/DotStatefulJob.java:12]: Codacy found an issue: [Field jobDetail has the same name as a method](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/main/java/com/dotmarketing/quartz/DotStatefulJob.java [hunk 8]
```diff
@@ -1,7 +1,31 @@
...
+        private String triggerGroupName;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/quartz/DotStatefulJob.java:13]: Codacy found an issue: [Field triggerGroupName has the same name as a method](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        assertTrue(DotStatefulTestJob.getThreadsExecuted() > 1);
+    }
+
+    /**
+     * Inner class with the logic to be executed by each thread in {@link
+     * DotStatefulJobTest#testDotStatefulJobIsReallyStateful()} ()}
+     */
+    private class StatefulTestThread implements Runnable {
+
+        private CyclicBarrier cyclicBarrier;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java:104]: Codacy found an issue: [Private field 'cyclicBarrier' could be made final; it is only initialized in the declaration or constructor.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/main/java/com/dotmarketing/quartz/DotJobStore.java [hunk 10]
```diff
@@ -16,9 +16,11 @@
 import org.quartz.JobPersistenceException;
 import org.quartz.ObjectAlreadyExistsException;
 import org.quartz.SchedulerConfigException;
+import org.quartz.StatefulJob;
 import org.quartz.Trigger;
 import org.quartz.core.SchedulingContext;
 import org.quartz.impl.jdbcjobstore.JobStoreCMT;
+import org.quartz.impl.jdbcjobstore.JobStoreSupport;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/quartz/DotJobStore.java:23]: Codacy found an issue: [Avoid unused imports such as 'org.quartz.impl.jdbcjobstore.JobStoreSupport'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+    /**
+     * Inner class with the logic to be executed by each thread in {@link
+     * DotStatefulJobTest#testTriggerBuilder()}
+     */
+    private class StatefulBuilderTestThread implements Runnable {
+
+        private CyclicBarrier cyclicBarrier;
+        private CountDownLatch countDownLatch;
+        private DotStatefulTestJob dotStatefulTestJob;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java:144]: Codacy found an issue: [Private field 'dotStatefulTestJob' could be made final; it is only initialized in the declaration or constructor.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+            this.countDownLatch = countDownLatch;
+            this.dotStatefulTestJob = dotStatefulTestJob;
+        }
+
+        @Override
+        public void run() {
+            final String thisThreadName = Thread.currentThread().getName();
+
+            try {
+                final JobDetail jd = getJobDetail(thisThreadName);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java:159]: Codacy found an issue: [Avoid variables with short names like jd](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

File: dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    /**
+     * Inner class with the logic to be executed by each thread in {@link
+     * DotStatefulJobTest#testDotStatefulJobIsReallyStateful()} ()}
+     */
+    private class StatefulTestThread implements Runnable {
+
+        private CyclicBarrier cyclicBarrier;
+        private CountDownLatch countDownLatch;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/quartz/DotStatefulJobTest.java:105]: Codacy found an issue: [Private field 'countDownLatch' could be made final; it is only initialized in the declaration or constructor.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5971105)

**Review by jdotcms — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/personas/business/DeleteMultiTreeUsedPersonaTagJob.java [hunk 14]
```diff
@@ -55,10 +55,8 @@ public static void triggerDeleteMultiTreeUsedPersonaTagJob(final User user,
         jobDetail.setVolatility(false);
         jobDetail.setRequestsRecovery(true);
 
-        final long startTime = System.currentTimeMillis();
-        final SimpleTrigger trigger = new SimpleTrigger("deleteMultitreeUnusedPersonasTrigger-" + randomID,
-                "delete_multitree_unusedpersonatag_triggers",
-                new Date(startTime));
+        final Trigger trigger = new TriggerBuilder().jobDetail(jobDetail)
+                .triggerGroupName("delete_multitree_unusedpersonatag_triggers").build();
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotmarketing/portlets/personas/business/DeleteMultiTreeUsedPersonaTagJob.java:59]: maybe it is better has a util for all this code

https://github.com/dotCMS/core/blob/2b0d9d698b39a8aa5247eb997fdcf4323e224930/dotCMS/src/main/java/com/dotmarketing/portlets/personas/business/DeleteMultiTreeUsedPersonaTagJob.java#L42

I don't know maybe a util method that receive the job class and job name, and the util method creted the job, trigger, etc

**Review by freddyucv — APPROVED** [review]

**Comment by nollymar** [comment]
This implementation won't be applied

**Pull request closed without merging**